### PR TITLE
protocols/noise: Inline `handshake` functions into upgrade traits

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

This fell out of https://github.com/libp2p/rust-libp2p/pull/2903. It removes a layer of indirection by inlining the various `handshake` functions into the respective upgrade traits. This should hopefully make it clearer, what message exchange actually happens for a particular handshake pattern.

This also removes a bit of complexity in terms of what the `{send,receive}_identity` functions do as several parts of it were conditional based on actually statically known state.

This is a breaking change because we make the entire `handshake` module private and thus prevent external users from re-using our handshake functions. However, I don't think having these public in the first place was a good move. The `libp2p-noise` crate specifically offers support for noise within libp2p and should not be used for noise handshakes in general.

Conveniently, this is also a net-negative in terms of code size :)

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- ~[ ] I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/noise/CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,12 @@
-# 0.39.1 [unreleased]
+# 0.40.0 [unreleased]
 
 - Introduce `NoiseAuthenticated::xx` constructor, assuming a X25519 DH key exchange. An XX key exchange and X25519 keys
   are the most common way of using noise in libp2p and thus deserve a convenience constructor. See [PR 2887].
+- Remove `Deref` implementation on `AuthenticKeypair`. See [PR XXXX].
+- Make `handshake` module private. See [PR XXXX].
 
 [PR 2887]: https://github.com/libp2p/rust-libp2p/pull/2887
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline transports/noise/CHANGELOG.md]:
```suggestion
- Remove `Deref` implementation on `AuthenticKeypair`. See [PR 2909].
- Make `handshake` module private. See [PR 2909].

[PR 2887]: https://github.com/libp2p/rust-libp2p/pull/2887
[PR 2909]: https://github.com/libp2p/rust-libp2p/pull/2909
```

**Comment by thomaseizinger** [comment]
@mxinden I removed the `Clone` of the keypair again in here by moving the functions to `ProtocolParams` as you suggested earlier. It actually works out well enough because the new function signature forces us to pass the prologue.

File: transports/noise/src/lib.rs [hunk 2]
```diff
@@ -249,24 +244,43 @@ where
...
     }
 }
 
-// Handshake pattern XX /////////////////////////////////////////////////////
-
+/// Implements the responder part of the `XX` noise handshake pattern.
+///
+/// `XX` is a 1.5 round-trip (3 messages) handshake.
+/// The first message in a noise handshake is unencrypted. In the `XX` handshake pattern, that message
+/// is empty and thus does not leak and information. The identities are then exchanged in the second
```
Comment (thomaseizinger) [inline transports/noise/src/lib.rs]:
```suggestion
/// is empty and thus does not leak any information. The identities are then exchanged in the second
```

File: transports/noise/src/lib.rs [hunk 3]
```diff
@@ -275,22 +289,44 @@ where
...
+        }
+        .boxed()
     }
 }
 
+/// Implements the initiator part of the `XX` noise handshake pattern.
+///
+/// `XX` is a 1.5 round-trip (3 messages) handshake.
+/// The first message in a noise handshake is unencrypted. In the `XX` handshake pattern, that message
+/// is empty and thus does not leak and information. The identities are then exchanged in the second
```
Comment (thomaseizinger) [inline transports/noise/src/lib.rs]:
```suggestion
/// is empty and thus does not leak any information. The identities are then exchanged in the second
```

**Comment by thomaseizinger** [comment]
@elenaf9 Do you have any idea why the autonat tests are failing on this branch? I can't figure it out :(

**Comment by elenaf9** [comment]
> @elenaf9 Do you have any idea why the autonat tests are failing on this branch? I can't figure it out :(

I don't think the test failures are autonat specific, but instead autonat is just the first package that is being tested. I just ran teh tests it locally and when I ignore all autonat tests, the next relevant test in line (`libp2p_core::tests::upgrade_pipeline`) fails as well.

**Comment by thomaseizinger** [comment]
> > @elenaf9 Do you have any idea why the autonat tests are failing on this branch? I can't figure it out :(
> 
> I don't think the test failures are autonat specific, but instead autonat is just the first package that is being tested. I just ran teh tests it locally and when I ignore all autonat tests, the next relevant test in line (`libp2p_core::tests::upgrade_pipeline`) fails as well.

Good thinking, I shall investigate!

**Comment by thomaseizinger** [comment]
`git bisect` saves the day once again. Our offending commit is ae18dd7ed600d0c692153b5a48d21d0442cf6e44 :)

**Comment by thomaseizinger** [comment]
This is ready for a full review @elenaf9 @mxinden !

**Review by mxinden — APPROVED** [review]
This looks good to me. Thanks for spotting the simplification!

**Pull request merged**